[GTK][WPE][Skia] Use GPU synchronization primitives for ImageBuffer/NativeImage during DisplayList recording/replaying

nikolaszimmermann opened the pull request:
#### 10435abe82ca31859fc6d3290c43857925f2df92
<pre>
[GTK][WPE][Skia] Use GPU synchronization primitives for ImageBuffer/NativeImage during DisplayList recording/replaying
<a href="https://bugs.webkit.org/show_bug.cgi?id=283292">https://bugs.webkit.org/show_bug.cgi?id=283292</a>

Reviewed by Said Abou-Hallawa.

Prepare for threaded GPU painting: Use new GPU synchronization
primitives in ImageBuffer/NativeImage during DisplayList
recording/replaying.

During recording, when a NativeImage/ImageBuffer is used call
finishAcceleratedRenderingAndCreateFence(), to make sure the
rendering commands are flushed to the GPU, and a fence is
injected in the GL command stream, so that we can wait during
replaying (in a worker thread!) that the GPU finished
processing, before we attempt to use the NativeImage or
ImageBuffer (in the accelerated case only!).

Covered by existing tests.

* Source/WebCore/platform/graphics/displaylists/DisplayList.h:
(WebCore::DisplayList::DisplayList::DisplayList):
(WebCore::DisplayList::DisplayList::replayOptions const):
* Source/WebCore/platform/graphics/displaylists/DisplayListItem.cpp:
(WebCore::DisplayList::applyFilteredImageBufferItem):
(WebCore::DisplayList::applyImageBufferItem):
(WebCore::DisplayList::applyNativeImageItem):
(WebCore::DisplayList::applySourceImageItem):
(WebCore::DisplayList::applySetStateItem):
(WebCore::DisplayList::applyItem):
* Source/WebCore/platform/graphics/displaylists/DisplayListItem.h:
* Source/WebCore/platform/graphics/displaylists/DisplayListRecorderImpl.cpp:
(WebCore::DisplayList::RecorderImpl::recordResourceUse):
* Source/WebCore/platform/graphics/displaylists/DisplayListReplayer.cpp:
(WebCore::DisplayList::Replayer::Replayer):
(WebCore::DisplayList::Replayer::replay):
* Source/WebCore/platform/graphics/displaylists/DisplayListReplayer.h:
(WebCore::DisplayList::Replayer::Replayer):
* Source/WebCore/platform/graphics/displaylists/DisplayListResourceHeap.h:
(WebCore::DisplayList::ResourceHeap::getImageBuffer):
(WebCore::DisplayList::ResourceHeap::getNativeImage):
(WebCore::DisplayList::ResourceHeap::getSourceImage):
(WebCore::DisplayList::ResourceHeap::getImageBuffer const): Deleted.
(WebCore::DisplayList::ResourceHeap::getNativeImage const): Deleted.
(WebCore::DisplayList::ResourceHeap::getSourceImage const): Deleted.
* Source/WebCore/platform/graphics/skia/SkiaThreadedPaintingPool.cpp:
(WebCore::SkiaThreadedPaintingPool::recordDisplayList const):

Canonical link: <a href="https://commits.webkit.org/286787@main">https://commits.webkit.org/286787@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e3ad558d09f7f24a0957336995ae3a00ae5fca9f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/77093 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/56128 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/30008 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/81649 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/28372 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/79210 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/65276 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/4424 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/60404 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/18469 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/80160 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/50365 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/66174 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/40711 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/47767 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/23672 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/26698 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/68882 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/24000 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/83078 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/4473 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/3005 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/68684 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/4628 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/66147 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/67938 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16945 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/11921 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/10011 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/4419 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/7235 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/4439 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/7874 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/6198 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

shallawa commented on Source/WebCore/platform/graphics/displaylists/DisplayListItem.h:
```diff
@@ -208,14 +208,19 @@ struct ApplyItemResult {
+enum class ApplyItemOptions : uint8_t {
+    Default,
+    FlushImagesAndWaitForCompletion
+};
```
I think this should be used as `OptionSet<ApplyItemOption>`.

```
enum class ApplyItemOption : uint8_t {
    FlushImagesAndWaitForCompletion = 1 << 0,
};
```

nikolaszimmermann replied:
Indeed, much nicer -- didn't think of `OptionSet`, d'oh!

shallawa commented on Source/WebCore/platform/graphics/displaylists/DisplayList.h:
```diff
@@ -65,9 +69,12 @@ class DisplayList {
 private:
     Vector<Item> m_items;
     ResourceHeap m_resourceHeap;
+    ApplyItemOptions m_options { ApplyItemOptions::Default };
```
This should be
```
OptionSet<ApplyItemOption> m_options;
```

shallawa commented on Source/WebCore/platform/graphics/displaylists/DisplayListItem.cpp:
```diff
@@ -51,34 +52,55 @@ bool isValid(const Item& item)
+        sourceImage->waitForAcceleratedRenderingFenceCompletion();
+#else
+    UNUSED_PARAM(options);
+#endif
```
I think this code should be in `ResourceHeap::getImageBuffer()`. So it will not be repeated in `applyImageBufferItem()`.

nikolaszimmermann replied:
Good idea!

shallawa commented on Source/WebCore/platform/graphics/displaylists/DisplayList.h:
```diff
@@ -44,6 +44,10 @@ class DisplayList {
 public:
     DisplayList() = default;
+    DisplayList(const ApplyItemOptions& options)
+        : m_options(options)
```
Which part of the code will pass these `ApplyItemOptions`? I will be great if this happens in this patch even if you pass the default option for SKIA for now.

nikolaszimmermann replied:
Great idea -- thanks @shallawa. PR #34502 contains the patch that makes use of the new `ApplyItemOptions` but I like your idea much more, to make it visible from the beginning where it's intended to be used.

Will adapt the PR and come up with a new version soon!

shallawa commented on Source/WebCore/platform/graphics/displaylists/DisplayListItem.h:
```diff
@@ -208,14 +208,18 @@ struct ApplyItemResult {
     std::optional<RenderingResourceIdentifier> resourceIdentifier;
 };
 
+enum class RecordReplayOption : uint8_t {
```
How about `ReplayOption`? The full name is `DisplayList::ReplayOption`

nikolaszimmermann replied:
Hmm, I thought about that, but it indicates the option would only affect replay, but it also affects recording.

shallawa commented on Source/WebCore/platform/graphics/displaylists/DisplayListResourceHeap.h:
```diff
@@ -92,23 +93,41 @@ class ResourceHeap {
     }
 
-    ImageBuffer* getImageBuffer(RenderingResourceIdentifier renderingResourceIdentifier) const
+    ImageBuffer* getImageBuffer(RenderingResourceIdentifier renderingResourceIdentifier, OptionSet<RecordReplayOption> options) const
```
`options` can be defaulted to `{ }` so non skia calls will not change.

nikolaszimmermann replied:
Ok, it was helpful during development, to assure I'm not misisng any callsite. Will adapt it now.

nikolaszimmermann replied:
Done for getSourceImage/getImageBuffer/getNativeImage.

shallawa commented on Source/WebCore/platform/graphics/displaylists/DisplayList.h:
```diff
@@ -65,9 +69,12 @@ class DisplayList {
     WEBCORE_EXPORT String asText(OptionSet<AsTextFlag>) const;
     void dump(WTF::TextStream&) const;
 
+    const OptionSet<RecordReplayOption>& applyItemOptions() const { return m_options; }
```
`applyItemOptions()` -> `replayOptions()`?

nikolaszimmermann replied:
Indeed, applyItemOptions() is a left-over... However since `recordAndReplayOptions()` sounds awkward, I'll go with your `ReplayOption` + `replayOptions()` suggestion.

shallawa commented on Source/WebCore/platform/graphics/skia/ImageBufferSkiaAcceleratedBackend.cpp:
```diff
@@ -108,6 +108,33 @@ ImageBufferSkiaAcceleratedBackend::~ImageBufferSkiaAcceleratedBackend()
+            grContext->submit(GrSyncCpu::kYes);
+    } else
+        grContext->flushAndSubmit(m_surface.get(), GrSyncCpu::kYes);
+}
```
I do not know much about skia but I think this function is doing two separate things. Can't it be split into two functions: one flushes the surface to the shared context and the other creates the fence?
```
flushToSharedDisplay()
createFence()
```

nikolaszimmermann replied:
We've followed the current pattern already in CoordinatedTileBuffer, which contains literally the same code related to flushAndSubmit + fence creation -- I've tried to stick to that. But yes, in principle we could split it -- however, fence creation might fail, and then we need to call again submit, but with the GrSyncCpu::kYes option to make the CPU block here if the GPU driver doesn't support fencing.

Therefore I think it's better to combine flush/submit+fencing into one call, that contains the logic in a single place.

shallawa commented on Source/WebCore/platform/graphics/skia/ImageBufferSkiaAcceleratedBackend.cpp:
```diff
@@ -108,6 +108,33 @@ ImageBufferSkiaAcceleratedBackend::~ImageBufferSkiaAcceleratedBackend()
+        grContext->flushAndSubmit(m_surface.get(), GrSyncCpu::kYes);
+}
+
+void ImageBufferSkiaAcceleratedBackend::waitForAcceleratedRenderingFenceCompletion()
```
Can't be just `waitForFenceCompletion()`?

`Accelerated` is already in the class name.

nikolaszimmermann replied:
Indeed, it's redundant, will adapt.

nikolaszimmermann replied:
Thought about this a second time: it's also part of the public ImageBuffer API, and there is no `Accelerated` in that class name. Currently the naming is symmetric:

`finishAcceleratedRenderingAndCreateFence()`
`waitForAcceleratedRenderingFenceCompletion()`

indicating that these two methods are used in combination.
I could rename one to `waitForFenceCompletion`, but not sure how to properly name the other - `finishRenderingAndCreateFence`?  I'm staying with the current names, unless you have a better suggestion :-)

shallawa commented on Source/WebCore/platform/graphics/skia/NativeImageSkia.cpp:
```diff
@@ -41,6 +40,35 @@ IGNORE_CLANG_WARNINGS_END
+    }
+
+    if (!m_fence)
+        grContext->submit(GrSyncCpu::kYes);
```
Why does not this function use `flushAndSubmit()` like what `ImageBufferSkiaAcceleratedBackend::finishAcceleratedRenderingAndCreateFence()` does?

nikolaszimmermann replied:
@alexgcastro asked the same -- Skias API is not orhogonal for SkSurface / SkImage -- the variant that we use for SkSurface doesn't exist for SkImage. You cannot pass the GrSyncCpu option to flushAndSubmit(SkImage) :-(
Therefore it has to be slightly different.

shallawa commented on Source/WebCore/platform/graphics/displaylists/DisplayListItem.cpp:
```diff
@@ -30,6 +30,7 @@
 #include "DisplayListResourceHeap.h"
 #include "FilterResults.h"
 #include "GraphicsContext.h"
+#include "NativeImage.h"
```
Is this header file needed?

nikolaszimmermann replied:
I will check again.

nikolaszimmermann replied:
It's no longer needed - it was needed to access the `backend()` method here -- now `DisplayList::ResourceHeap` is responsible, so this can go away.

shallawa commented on Source/WebCore/platform/graphics/displaylists/DisplayList.h:
```diff
@@ -44,6 +44,10 @@ class DisplayList {
+    explicit DisplayList(OptionSet<RecordReplayOption> options)
+        : m_options(options)
+    {
+    }
```
I think if you default `options` to `{ }` you can remove the default constructor.

nikolaszimmermann replied:
Makes sense, will adapt.

nikolaszimmermann committed:
[GTK][WPE][Skia] Use GPU synchronization primitives for ImageBuffer/NativeImage during DisplayList recording/replaying
https://bugs.webkit.org/show_bug.cgi?id=283292

Reviewed by Said Abou-Hallawa.

Prepare for threaded GPU painting: Use new GPU synchronization
primitives in ImageBuffer/NativeImage during DisplayList
recording/replaying.

During recording, when a NativeImage/ImageBuffer is used call
finishAcceleratedRenderingAndCreateFence(), to make sure the
rendering commands are flushed to the GPU, and a fence is
injected in the GL command stream, so that we can wait during
replaying (in a worker thread!) that the GPU finished
processing, before we attempt to use the NativeImage or
ImageBuffer (in the accelerated case only!).

Covered by existing tests.

* Source/WebCore/platform/graphics/displaylists/DisplayList.h:
(WebCore::DisplayList::DisplayList::DisplayList):
(WebCore::DisplayList::DisplayList::replayOptions const):
* Source/WebCore/platform/graphics/displaylists/DisplayListItem.cpp:
(WebCore::DisplayList::applyFilteredImageBufferItem):
(WebCore::DisplayList::applyImageBufferItem):
(WebCore::DisplayList::applyNativeImageItem):
(WebCore::DisplayList::applySourceImageItem):
(WebCore::DisplayList::applySetStateItem):
(WebCore::DisplayList::applyItem):
* Source/WebCore/platform/graphics/displaylists/DisplayListItem.h:
* Source/WebCore/platform/graphics/displaylists/DisplayListRecorderImpl.cpp:
(WebCore::DisplayList::RecorderImpl::recordResourceUse):
* Source/WebCore/platform/graphics/displaylists/DisplayListReplayer.cpp:
(WebCore::DisplayList::Replayer::Replayer):
(WebCore::DisplayList::Replayer::replay):
* Source/WebCore/platform/graphics/displaylists/DisplayListReplayer.h:
(WebCore::DisplayList::Replayer::Replayer):
* Source/WebCore/platform/graphics/displaylists/DisplayListResourceHeap.h:
(WebCore::DisplayList::ResourceHeap::getImageBuffer):
(WebCore::DisplayList::ResourceHeap::getNativeImage):
(WebCore::DisplayList::ResourceHeap::getSourceImage):
(WebCore::DisplayList::ResourceHeap::getImageBuffer const): Deleted.
(WebCore::DisplayList::ResourceHeap::getNativeImage const): Deleted.
(WebCore::DisplayList::ResourceHeap::getSourceImage const): Deleted.
* Source/WebCore/platform/graphics/skia/SkiaThreadedPaintingPool.cpp:
(WebCore::SkiaThreadedPaintingPool::recordDisplayList const):

Canonical link: https://commits.webkit.org/286787@main
